Sources/NIOHPACK/HuffmanCoding.swift

```diff
@@ -0,0 +1,221 @@
+            } else {
+                state.remainingBits = 8 - nbits
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code)
+            }
+        }
+    }
+    
+    private mutating func ensureBitsAvailable(_ bits: Int, _ buffer: inout ByteBuffer) {
+        let bytesNeeded = bits / 8
+        if bytesNeeded <= buffer.writableBytes {
+            // just zero the requested number of bytes before we start OR-ing in our values
+            buffer.withUnsafeMutableWritableBytes { ptr in
+                ptr.baseAddress!.assumingMemoryBound(to: UInt8.self).assign(repeating: 0, count: bytesNeeded)
+            }
+            return
+        }
+        
+        // one extra byte to ensure we have space to fill at the end of the encoded string, if we need it
+        let neededToAdd = bytesNeeded - buffer.writableBytes
+        let newLength = buffer.capacity + neededToAdd
+        
+        // reallocate the buffer to ensure we have the room we need
+        buffer.changeCapacity(to: newLength)
+        
+        // now zero all writable bytes that we expect to use
+        buffer.withUnsafeMutableWritableBytes { ptr in
+            ptr.baseAddress!.assumingMemoryBound(to: UInt8.self).assign(repeating: 0, count: bytesNeeded)
+        }
+    }
+}
+
+/// Errors that may be encountered by the Huffman decoder.
+public enum HuffmanDecoderError : Error
+{
+    /// The decoder entered an invalid state. Usually this means invalid input.
+    case invalidState
+    /// The output data could not be validated as UTF-8.
+    case decodeFailed
+}
+
+/// A `HuffmanDecoder` is an idempotent class that performs string decoding using
+/// a static table of values. It maintains no internal state, so is thread-safe.
+public struct HuffmanDecoder {
+    /// The decoder table. This structure doesn't actually take up any space, I think?
+    let decoderTable = HuffmanDecoderTable()
+    
+    /// Initializes a new Huffman decoder.
+    public init() {}
+    
+    /// Decodes a single string from a view into a `ByteBuffer`.
+    ///
+    /// - Parameter input: A `ByteBufferView` over the entire set of bytes
+    ///                     that comprise the encoded string.
+    /// - Parameter output: A `ByteBuffer` into which decoded UTF-8 octets will be
+    ///                     written.
+    /// - Returns: The number of UTF-8 characters written into the output `ByteBuffer`.
+    /// - Throws: HuffmanDecoderError if the data could not be decoded.
+    @discardableResult
+    public func decodeString(from input: ByteBufferView, into output: inout ByteBuffer) throws -> Int {
+        if input.count == 0 {
+            return 0
+        }
+        
+        var state: UInt8 = 0
+        var acceptable = false
+        
+        // we write to an intermediate contiguous array, because write(integer:) goes through
+        // all sorts even when writing individual bytes.
```

Comment:
Surely it's not the case that allocating a new buffer, writing to that, and then copying into the ByteBuffer is faster than using `write(integer:)`?

Reply:
It appeared that way when I ran the performance tests. Performance when I moved to using `write(integer:)` dropped like a stone…

Reply:
My suspicion was that it had something to do with the fact that it was doing everything via pointers, even with a single byte that could just be written individually.

Reply:
…and also doing one-byte bounds checking on every single byte. It caused a roughly 600% drop in performance.

Reply:
I've updated the unit tests to run the performance tests (large blobs of string data) and when using `write(integer:)` I'm seeing and average of ~0.75 seconds, while using a `ContiguousArray` and a single `write(bytes:)` I get ~0.3 seconds:

Using `write(integer:)`:

`Case '-[NIOHPACKTests.HuffmanCodingTests testBasicDecodingPerformanceWithByteWrites]' measured [Time, seconds] average: 0.741, relative standard deviation: 9.097%, values: [0.940713, 0.707361, 0.747987, 0.705596, 0.725697, 0.723977, 0.713024, 0.709017, 0.716341, 0.725194], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "", baselineAverage: , maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100`

Using `write(bytes:)`:

`Case '-[NIOHPACKTests.HuffmanCodingTests testBasicDecodingPerformance]' measured [Time, seconds] average: 0.299, relative standard deviation: 1.638%, values: [0.304642, 0.295964, 0.301061, 0.305507, 0.294101, 0.295073, 0.303937, 0.292762, 0.292673, 0.302078], performanceMetricID:com.apple.XCTPerformanceMetric_WallClockTime, baselineName: "Local Baseline", baselineAverage: 0.318, maxPercentRegression: 10.000%, maxPercentRelativeStandardDeviation: 10.000%, maxRegression: 0.100, maxStandardDeviation: 0.100`

Reply:
Other tests:

Using `write(bytes:)`:

- Test Case '-[NIOHPACKTests.HuffmanCodingTests testBasicCoding]' passed (0.274 seconds).
- Test Case '-[NIOHPACKTests.HuffmanCodingTests testComplexCoding]' passed (0.254 seconds).

Using `write(integer:)`:

- Test Case '-[NIOHPACKTests.HuffmanCodingTests testBasicCoding]' passed (0.224 seconds).
Test Case '-[NIOHPACKTests.HuffmanCodingTests testComplexCoding]' passed (0.231 seconds).

So, there's an argument to switch from one method to the other at some input size. I suppose the question is when we'll encounter inputs of that size?

Reply:
Er, actually, those results should be the other way around, I think. I wrote a test to explicitly decode larger and larger buffers using both styles, and the `ContiguousArray` version generally runs in ~40% of the time of the `write(integer:)` version:

Test Case '-[NIOHPACKTests.HuffmanCodingTests testCompareDecodingStrategies]' started.
Iteration 0: Decoded octets = 245, Integers = 0.0012660026550292969s, Bytes = 0.0005530118942260742s, Ratio = 0.43681732580037663
Iteration 1: Decoded octets = 490, Integers = 0.0034099817276000977s, Bytes = 0.0015499591827392578s, Ratio = 0.45453592029365497
Iteration 2: Decoded octets = 735, Integers = 0.00506901741027832s, Bytes = 0.002640962600708008s, Ratio = 0.5210008936550492

…

Iteration 60: Decoded octets = 14945, Integers = 0.04639601707458496s, Bytes = 0.021033048629760742s, Ratio = 0.4533373758344082
Iteration 61: Decoded octets = 15190, Integers = 0.05168294906616211s, Bytes = 0.02321600914001465s, Ratio = 0.44920054988144337
Iteration 62: Decoded octets = 15435, Integers = 0.04749500751495361s, Bytes = 0.01870596408843994s, Ratio = 0.39385116598940306
Iteration 63: Decoded octets = 15680, Integers = 0.04568898677825928s, Bytes = 0.019114017486572266s, Ratio = 0.41835065372181796
Iteration 64: Decoded octets = 15925, Integers = 0.0474470853805542s, Bytes = 0.0235140323638916s, Ratio = 0.49558433727372087
Average ratio: 0.4201896264685576

Reply:
Can I confirm you're using optimised builds in both cases?